### ISSUE TITLE
Unable to verify code in KC 23

### ISSUE THREAD

**Issue by finalgamer** [RedFroggy/keycloak-verify-email-by-code#17]
We have this issue since upgrading to keycloak v23.

The required action form is loaded correctly. After entering the code sent to the email an "Internal Server Error" happens.
I am unable to find a lot of logs to this.

The `Internal Server Error` happens at `POST /realms/nau/login-actions/required-action`

## Logs
`
[org.keycloak.events] (executor-thread-2) type=REGISTER_ERROR, realmId=8d347b8a-0382-48c8-a443-629f6963bd21, clientId=account-console, userId=null, ipAddress=CENSORED, error=invalid_registration, auth_method=openid-connect, auth_type=code, register_method=form, redirect_uri=CENSORED/realms/nau/account/, code_id=6efe42c9-d168-4642-89df-e6b8f97521d7, email=michele@nau.ch, username=michele@nau.ch
`

## Event Logs
CUSTOM_REQUIRED_ACTION_ERROR
`expired_code`

I am honestly not sure how I can debug this further. So looking for help here.

**Comment by finalgamer** [comment]
Just ran a test with a minimal setup running the docker-compose.yaml provided in this repository with Keycloak 23.

Here is the error from the console

```
keycloak-verify-email-by-code-keycloak-1  | 2024-01-18 12:33:56,760 DEBUG [org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext] (executor-thread-11) Restarting handler chain for exception exception: java.lang.NullPointerException: Cannot invoke "org.jboss.resteasy.spi.HttpRequest.getDecodedFormParameters()" because the return value of "org.keycloak.common.util.ResteasyProvider.getContextData(java.lang.Class)" is null
keycloak-verify-email-by-code-keycloak-1  |     at fr.redfroggy.keycloak.requiredactions.VerifyEmailByCode.processAction(VerifyEmailByCode.java:109)
keycloak-verify-email-by-code-keycloak-1  |     at org.keycloak.services.resources.LoginActionsService.processRequireAction(LoginActionsService.java:1090)
keycloak-verify-email-by-code-keycloak-1  |     at org.keycloak.services.resources.LoginActionsService.requiredActionPOST(LoginActionsService.java:1025)
keycloak-verify-email-by-code-keycloak-1  |     at org.keycloak.services.resources.LoginActionsService$quarkusrestinvoker$requiredActionPOST_677a8efd4e80bfe1b3aa5a0d6fca2043252c9624.invoke(Unknown Source)
keycloak-verify-email-by-code-keycloak-1  |     at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
keycloak-verify-email-by-code-keycloak-1  |     at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)
keycloak-verify-email-by-code-keycloak-1  |     at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:145)
keycloak-verify-email-by-code-keycloak-1  |     at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:576)
keycloak-verify-email-by-code-keycloak-1  |     at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
keycloak-verify-email-by-code-keycloak-1  |     at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
keycloak-verify-email-by-code-keycloak-1  |     at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
keycloak-verify-email-by-code-keycloak-1  |     at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
keycloak-verify-email-by-code-keycloak-1  |     at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
keycloak-verify-email-by-code-keycloak-1  |     at java.base/java.lang.Thread.run(Thread.java:840)
keycloak-verify-email-by-code-keycloak-1  | 
keycloak-verify-email-by-code-keycloak-1  | 2024-01-18 12:33:56,761 ERROR [org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-11) Uncaught server error: java.lang.NullPointerException: Cannot invoke "org.jboss.resteasy.spi.HttpRequest.getDecodedFormParameters()" because the return value of "org.keycloak.common.util.ResteasyProvider.getContextData(java.lang.Class)" is null
keycloak-verify-email-by-code-keycloak-1  |     at fr.redfroggy.keycloak.requiredactions.VerifyEmailByCode.processAction(VerifyEmailByCode.java:109)
keycloak-verify-email-by-code-keycloak-1  |     at org.keycloak.services.resources.LoginActionsService.processRequireAction(LoginActionsService.java:1090)
keycloak-verify-email-by-code-keycloak-1  |     at org.keycloak.services.resources.LoginActionsService.requiredActionPOST(LoginActionsService.java:1025)
keycloak-verify-email-by-code-keycloak-1  |     at org.keycloak.services.resources.LoginActionsService$quarkusrestinvoker$requiredActionPOST_677a8efd4e80bfe1b3aa5a0d6fca2043252c9624.invoke(Unknown Source)
keycloak-verify-email-by-code-keycloak-1  |     at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
keycloak-verify-email-by-code-keycloak-1  |     at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)
keycloak-verify-email-by-code-keycloak-1  |     at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:145)
keycloak-verify-email-by-code-keycloak-1  |     at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:576)
keycloak-verify-email-by-code-keycloak-1  |     at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
keycloak-verify-email-by-code-keycloak-1  |     at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
keycloak-verify-email-by-code-keycloak-1  |     at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
keycloak-verify-email-by-code-keycloak-1  |     at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
keycloak-verify-email-by-code-keycloak-1  |     at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
keycloak-verify-email-by-code-keycloak-1  |     at java.base/java.lang.Thread.run(Thread.java:840)
```

**Comment by finalgamer** [comment]
Okay, I am in no way a java developer and I have no idea what I am doing. But I got it working.

I have replaced this line
https://github.com/RedFroggy/keycloak-verify-email-by-code/blob/f162c2b5e8193afe2fb2a36a5756b63e896825ab/src/main/java/fr/redfroggy/keycloak/requiredactions/VerifyEmailByCode.java#L109

With this

```java
KeycloakSession session = Resteasy.getContextData(KeycloakSession.class);
MultivaluedMap<String, String> formData = session.getContext().getHttpRequest().getDecodedFormParameters();
```

I found out about the change in the Keycloak v23 changelogs. Specifically this issue https://github.com/keycloak/keycloak/issues/10713

and then this commit and file change
https://github.com/keycloak/keycloak/pull/15659/files#diff-77e7630f0333c606e0052b7849288a9d1a6a9b6d2c3958e266ab5d01c7495e0e


I hope this information helps you to update the package as I have no clue how to update the tests and why this is working.

**Comment by danieliu456** [comment]
Having the same issue @dajay is there any chance we can expect an upgrade for this provider?

**Comment by SangI762** [comment]
Same for us, upgraded keycloak to v23 and got this error.

**Comment by dajay** [comment]
Hello every body, fix In progress

**Comment by dajay** [comment]
Hello,
New release is available [23.0.0](https://github.com/RedFroggy/keycloak-verify-email-by-code/releases/tag/23.0.0).
Thanks
